src/runtime_src/core/include/xgq.h

```diff
@@ -215,4 +232,284 @@ struct xrt_cmd_exit_ert {
 	};
 };
 
+/*
+ * Generic XGQ implementation.
+ *
+ * One XGQ consists of one submission (SQ) and one completion ring (CQ) buffer shared by one client
+ * and one server. Client send request through SQ to server, which processes it and send back
+ * response through CQ.
+ */
+
+#define XGQ_ALLOC_MAGIC			0x5847513F	/* XGQ? */
+#define XGQ_ATTACH_MAGIC		0x58475121	/* XGQ! */
+#define XGQ_MAJOR			1
+#define XGQ_MINOR			0
+#define XGQ_MIN_NUM_SLOTS		4
+#define XGQ_RING_LEN(nslots, slotsz)	\
+	(sizeof(struct xgq_header) + (nslots) * ((slotsz) + sizeof(struct xrt_com_queue_entry)))
+
+/*
+ * Meta data shared b/w client and server of XGQ
+ */
+struct xgq_header {
+	uint32_t xh_magic; /* Always the first member */
+
+	uint32_t xh_minor:8;
+	uint32_t xh_major:8;
+	uint32_t xh_rsvd:16;
+
+	/* SQ and CQ share the same num of slots. */
+	uint32_t xh_slot_num;
+
+	uint32_t xh_sq_offset;
+	uint32_t xh_sq_slot_size;
+	uint32_t xh_cq_offset;
+	/* CQ slot size and format is tied to XGQ version. */
+
+	/*
+	 * Consumed pointer for both SQ and CQ are here since they don't generate interrupt,
+	 * so no need to be a register.
+	 */
+	uint32_t xh_sq_consumed;
+	uint32_t xh_cq_consumed;
+};
+
+/* Software representation of a single ring buffer. */
+struct xgq_ring {
+	uint32_t xr_slot_num;
+	uint32_t xr_slot_sz;
+	uint32_t xr_produced;
+	uint32_t xr_consumed;
+
+	uint64_t xr_produced_addr;
+	uint64_t xr_consumed_addr;
+	uint64_t xr_slot_addr;
+};
+
+/* Software representation of a single XGQ. */
+struct xgq {
+	bool xq_is_server;
+	uint64_t io_hdl;
+	struct xgq_ring xq_sq ____cacheline_aligned_in_smp;
+	struct xgq_ring xq_cq ____cacheline_aligned_in_smp;
+};
+
+
+/*
+ * XGQ implementation details and helper routines.
+ */
+
+static inline void xgq_copy_to_ring(uint64_t io_hdl, void *buf, uint64_t tgt, size_t len)
+{
+	size_t i;
+	uint32_t *src = (uint32_t *)buf;
+
+	for (i = 0; i < len / 4; i++, tgt += 4)
+		xgq_mem_write32(io_hdl, tgt, src[i]);
+}
+
+static inline void xgq_copy_from_ring(uint64_t io_hdl, void *buf, uint64_t src, size_t len)
+{
+	size_t i;
+	uint32_t *tgt = (uint32_t *)buf;
+
+	for (i = 0; i < len / 4; i++, src += 4)
+		tgt[i] = xgq_mem_read32(io_hdl, src);
```

Comment:
This will not compile since xgq_mem_read32 is not visible by a compilation unit that includes just this header file.
The function appears to be defined in xgq_hwemu_plat.h but that is a indirect dependency that shouldn't be there.

Reply:
The idea here is to abstract a generic XGQ implementation which can be shared by all platforms using XGQ. For platform specific operations like xgq_mem_read32, they need to be defined in a platform specific header file that is included before this generic header.

Any suggestion we can achieve this while xgq.h can be compiled by itself? Or can we make some exception to skip this check for xgq.h?

Reply:
can we do #ifndef xgq_mem_read32 then define it as a empty function inside xgq.h? Also put some comment here to say that this is just for this check. it won't work when you really run the code for sure.

Reply:
Hmm, this won't work since inside the xgq_xxx_platform.h, functions like xgq_mem_read32 is inline function instead of macro. And we'd better keep them as inline functions.

Reply:
Made the changes as discussed.
1) split xgq.h to xgq.h and xgq_impl.h
2) add XGQ_IMPL macro and include xgq_impl.h in xgq_hwemu_plat.h
3) add virtual platform specific definition in xgq_impl.h if XGQ_IMPL is not defined.

@maxzhen  and @stsoe  please take a look again. Thanks!